### ISSUE TITLE
Generating references from files in S3 (using kerchunk + fsspec)

### ISSUE THREAD

**Issue by TomNicholas** [TomNicholas/VirtualiZarr#61]
Raised by @abarciauskas-bgse in https://github.com/TomNicholas/VirtualiZarr/issues/60.

Presumably kerchunk's backends opener functions can already do this? In which case we just need to pass the right kwargs / fsspec whatevers to those inside `open_virtual_dataset`.

**Comment by norlandrhagen** [comment]
Wondering how we want to package up kwargs for all the different kerchunk file readers so we can pipe in things like `storage_options` to get s3 reading to work.

netcdf3, netcdf4, tiff etc. all have a few options. ex:

```
SingleHdf5ToZarr:

  h5f: 'BinaryIO | str | h5py.File | h5py.Group',
  url: str = None,
  spec=1,
  inline_threshold=500,
  storage_options=None,
  error='warn',
  vlen_encode='embed',
  out=None,


NetCDF3ToZarr:

def __init__(
    self,
    filename,
    storage_options=None,
    inline_threshold=100,
    max_chunk_size=0,
    out=None,
    **kwargs)

tiff_to_zarr:

    urlpath: str
        Location of input TIFF
    remote_options: dict
        pass these to fsspec when opening urlpath
    target: str
        Write JSON to this location. If not given, no file is output
    target_options: dict
        pass these to fsspec when opening target

etc..
```

**Comment by TomNicholas** [comment]
Urgh that's messy - I'm not familiar at all with these so would defer to your judgement. But generally they should all go in one dict so that they can be cleanly separated from Xarray kwargs once open_virtual_dataset is replaced by xr.open_dataset(..., engine='virtualizarr').

**Comment by TomNicholas** [comment]
Whilst I think the approach in #67 is fine for now, is it possible to imagine any other implementation of reading byte ranges from remote files that doesn't rely on fsspec / kerchunk?

**Comment by rabernat** [comment]
> is it possible to imagine any other implementation of reading byte ranges from remote files that doesn't rely on fsspec / kerchunk?

It's straightforward to abstract this via a wrapper class. Just imagine the API you'd like to use from the virtualizarr perspective, create an ABC interface for it, and then make an fsspec implementation.

**Comment by norlandrhagen** [comment]
@TomNicholas that would be nice! For a stop-gap, happy to continue working on #67 or if you want to pause and look for other file reader paths sounds good to.

**Comment by abarciauskas-bgse** [comment]
@TomNicholas curious about this comment https://github.com/TomNicholas/VirtualiZarr/issues/61#issuecomment-2027723156

>is it possible to imagine any other implementation of reading byte ranges from remote files that doesn't rely on fsspec / kerchunk?

since the current implementation relies on reading files using kerchunk - are you thinking about changing that implementation entirely? 

If keeping the current implementation, which relies on kerchunk modules, I think #67 makes sense as that PR just builds on that integration via the additional parameters kerchunk needs to open remote files.

**Comment by abarciauskas-bgse** [comment]
>IIUC with zarr v3 and the rust object-store create we also wouldn't need fsspec to actually read data from disk/object storage

I would say it's ok to postpone S3 support until some of those things you mentioned (zarr chunk manifests, zarr v3 rust object-store) exists - wdyt @norlandrhagen 

Is there already WIP `rust object-store` @TomNicholas

**Comment by TomNicholas** [comment]
The [rust crate exists](https://crates.io/crates/object_store/0.3.0), but you'll have to ask @jhamman about the status of attempts to use it to load data into python.

**Comment by TomNicholas** [comment]
@abarciauskas-bgse actually see https://github.com/zarr-developers/zarr-python/pull/1661

**Comment by TomNicholas** [comment]
So the rust `object-store` crate looks like a potential solution to the problem of reading bytes from known byte ranges from S3 without using fsspec. That's great for reading from zarr stores created by virtualizarr.

But to actually use `object-store` to read the netCDF file to determine the correct byte ranges to put in the manifest is a different requirement. That requires re-creating the [logic in `kerchunk.hdf.SingleHdfToZarr`](https://github.com/fsspec/kerchunk/blob/a0c4f3b828d37f6d07995925b324595af68c4a19/kerchunk/hdf.py#L93), which uses `h5py` and `fsspec` like this:

```python
if isinstance(h5f, str):
    fs, path = fsspec.core.url_to_fs(h5f, **(storage_options or {}))
    input_file = fs.open(path, "rb")
    url = h5f
    _h5f = h5py.File(input_file, mode="r")
```

Is there any way to do _this_ without using fsspec?

**Comment by TomNicholas** [comment]
Answering my own question, how about this `ro3` file open mode on the h5py docs? Says it enables read-only access to HDF5 files in the AWS S3 or S3-compatible object stores. Then we could implement our own S3-compatible netCDF opener using `h5py` but not `fsspec`...?

https://docs.h5py.org/en/stable/high/file.html#file-drivers

**Comment by TomNicholas** [comment]
> Is this within scope for VirtualiZarr right now?

I agree it's fairly orthogonal, and not part of an MVP (which would rely on fsspec and kerchunk both for generating references and reading actual bytes later). But if we're thinking about how to create references that don't require fsspec to read bytes from (i.e. zarr stores with chunk manifests), then it's also interesting to think about how to generate those references without fsspec. 

But there is less of an obvious motivation for doing so other than reliability / performance / design simplicity (whereas zarr stores without fsspec get us an actual feature: language-agnostic data access).

**Comment by TomNicholas** [comment]
I think we've gone off track from the original issue here, so I've opened #78 to specifically discuss generating references from data in S3 without using fsspec (as opposed to reading bytes from known references to data in S3).

I'll leave this issue to discuss the generating of references from files in S3 using kerchunk + fsspec.

**Comment by TomNicholas** [comment]
@abarciauskas-bgse @norlandrhagen it would be great to try to move forward with using kerchunk + fsspec for now (i.e. PR #67) with the understanding that eventually we would like to replace those dependencies with the ideas in #78.

**Comment by TomNicholas** [comment]
I think this was closed by @norlandrhagen in #67, with the other ideas being tracked in #78.
